File: Sesion 10/Ejercicio/Ejercicio3 S10/ejercicio3.py
from typing import List, TypeVar

T = TypeVar('T')

class Recipiente:
    def __init__(self, arr: List[T]):
        self.arreglo = arr

    def imprimir_arreglo(self) -> None:
        for elemento in self.arreglo:
            print(elemento, end=" ")
        print()

    def ordenar_arreglo(self) -> None:
        numero_total = len(self.arreglo)
        for i in range(numero_total - 1):
            indice_menor = i
            for j in range(i + 1, numero_total):
                if self.arreglo[j] < self.arreglo[indice_menor]:
                    indice_menor = j
            if indice_menor != i:
                self.arreglo[i], self.arreglo[indice_menor] = self.arreglo[indice_menor], self.arreglo[i]


arreglo_enteros = [5, 2, 7, 3, 1, 9]
recipiente_enteros = Recipiente(arreglo_enteros)

print("\nArreglo de enteros original:")
recipiente_enteros.imprimir_arreglo()

print("\nArreglo de enteros ordenado:")
recipiente_enteros.ordenar_arreglo()
recipiente_enteros.imprimir_arreglo()

arreglo_dobles = ["Siempre buenos dias", "Ingeniero", "Oscar", "Pongame", "20", "porfavor"]
recipiente_dobles = Recipiente(arreglo_dobles)

print("\nArreglo de strings original:")
recipiente_dobles.imprimir_arreglo()

print("\nArreglo de strings ordenado:")
recipiente_dobles.ordenar_arreglo()
recipiente_dobles.imprimir_arreglo()

File: Sesion 10/Ejercicio/Ejercicio3 S10/ejercicio3.cpp
#include <iostream>
#include <vector>
using namespace std;

template <typename T>
class Recipiente {
    private:
        vector<T> arreglo;
    public:
        Recipiente(const vector<T>& arr) : arreglo(arr) {}

        void imprimirArreglo() const {
            for (auto elemento : arreglo) {
                cout << elemento << " ";
            }
            cout << endl;
        }

        void ordenarArreglo() {
            int numero_total = arreglo.size();
            for (int i = 0; i < numero_total - 1; ++i) {
                int indice_menor = i;
                for (int j = i + 1; j < numero_total; ++j) {
                    if (arreglo[j] < arreglo[indice_menor]) {
                        indice_menor = j;
                    }
                }
                if (indice_menor != i) {
                    swap(arreglo[i], arreglo[indice_menor]);
                }
            }
        }
};

int main() {
    vector<int> arreglo_enteros = {5, 2, 7, 3, 1, 9};
    Recipiente<int> recipiente_enteros(arreglo_enteros);

    cout << "\nArreglo de enteros original:" << endl;
    recipiente_enteros.imprimirArreglo();

    cout << "\nArreglo de enteros ordenado:" << endl;
    recipiente_enteros.ordenarArreglo();
    recipiente_enteros.imprimirArreglo();

    vector<string> arreglo_dobles = {"Siempre buenos dias", "Ingeniero", "Oscar", "Pongame", "20", "porfavor"}; 
    Recipiente<string> recipiente_dobles(arreglo_dobles);

    cout << "\nArreglo de strings original:" << endl;
    recipiente_dobles.imprimirArreglo();

    cout << "\nArreglo de strings ordenado:" << endl;
    recipiente_dobles.ordenarArreglo(); // Compara los elementos basándose en sus valores ASCII y realiza intercambios para colocar los elementos en orden creciente.
    recipiente_dobles.imprimirArreglo();

    return 0;
}
